contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,364 @@
+    lastPrices[productId][address(stakingPool)] = basePrice;
+    lastPriceUpdate[productId][address(stakingPool)] = block.timestamp;
+
+    stakingPool.buyCover(
+      productId,
+      coveredAmount,
+      REWARD_BPS * price / BASIS_PRECISION,
+      period,
+      capacityFactor
+    );
+
+    stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+
+    return (coveredAmount, price);
+  }
+
+  function extendCover(
+    uint coverId,
+    uint32 period,
+    uint96 amount,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external returns (uint) {
+    require(_isApprovedOrOwner(_msgSender(), coverId), "Cover: caller is not owner nor approved");
+
+    Cover memory cover = covers[coverId];
+    (uint newCoverId, uint priceInAsset) = _createCover(
+      ERC721.ownerOf(coverId),
+      cover.productId,
+      cover.payoutAsset,
+      amount,
+      period,
+      stakingPools
+    );
+
+    // make the cover expire at current block
+    uint32 newPeriod = uint32(block.timestamp) - cover.start;
+    uint32 previousPeriod = covers[coverId].period;
+    uint priceAlreadyPaid = (previousPeriod - newPeriod) / previousPeriod * cover.price;
+    covers[coverId].period = newPeriod;
+
+    if (priceInAsset > priceAlreadyPaid) {
+      // get price for already paid asset
+      uint priceToBePaid = priceInAsset - priceAlreadyPaid;
+      require(priceToBePaid <= maxPrice, "Cover: Price exceeds maxPrice");
+      retrievePayment(priceToBePaid, cover.payoutAsset);
+    }
+
+    return newCoverId;
+  }
+
+  function addAmount(
+    uint coverId,
+    uint96 amount,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external {
+    Cover memory cover = covers[coverId];
+
+    uint32 period = uint32(block.timestamp) - cover.start;
+    // convert to NXM amount
+    uint amountToCover = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(cover.payoutAsset));
+    uint totalPremiumInNXM = 0;
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountToCover == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint price) = buyCoverFromPool(stakingPool, cover.productId, amountToCover, period);
+      amountToCover -= coveredAmount;
+      totalPremiumInNXM += price;
+      stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+    }
+    require(amountToCover == 0, "Not enough available capacity");
+
+    uint premiumInAsset = totalPremiumInNXM * pool().getTokenPrice(pool().assets(cover.payoutAsset)) / 1e18;
+
+    require(premiumInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(premiumInAsset, cover.payoutAsset);
+
+    cover.amount += amount;
+    cover.price += uint96(premiumInAsset);
+  }
+
+  function addPeriod(uint coverId, uint32 extraPeriod, uint maxPrice) external {
```

Comment:
what is maxPrice? looks like an absolute value in NXM from the implementation - doesn't look like what we want